### DISCUSSION TITLE
Celery not propagating error message to results backend when raising werkzeug HTTP Exceptions.

### DISCUSSION THREAD

**Discussion by amthorn** [celery/celery#7159] [Q&A]
I am building a simple flask application using celery to facilitate micro-service intercommunication. In the process of doing this, i am raising HTTP exceptions inside of various celery tasks which are then propagated up to flask and raised appropriately to the flask engine (Which avoids having to catch and re-raise an exception).

However, the exception message is not being propagated to the result backend. Through investigation, i believe this is because of the following two conditions:

1. Celery serializes the error message to the result backend by serializing the "args" attribute on the error message which is raised. This is thus not being propagated to the exc_message key and is thus not being propagated into the results backend
See Code: https://github.com/celery/celery/blob/master/celery/backends/base.py#L336

2. Werkzeug HTTP exceptions do not implement an "args" attribute on any of their exceptions, instead opting to use "description" and "body" attributes for those messages. Attempting to access the "args" attribute will always return an empty tuple.
See Code: https://github.com/pallets/werkzeug/blob/main/src/werkzeug/exceptions.py

This means that Celery will not capture the error message string for these errors and will default to using no error message and the actual message raised from inside the task will be lost.

There are two possible options:
1. Celery team agrees to only supports exceptions that contain an "args" attribute which contains all the relevant exception information.
2. Celery is updated in the backend base file to use "str()" to extract the error message (possibly in addition to args as a separate key).

I tend to prefer (2) for several reasons:

1. This is much more common of an approach to retrieving the error message from an exception object and is also implemented by a broader range of custom exceptions that exist in python libraries. This is also how the python engine actually propogates the error message to stdout and can thus be relied upon by celery.

2. Limiting celery's support to only certain kinds of error messages being propogated should be well-documented. When an error is or is not supported through error propagation should be noted so users can search to find this information somewhere. This would lead to a more difficult time using celery.

Is this a bug? Can anyone else reproduce this issue? And should it be fixed?

**NOTE**
I do have a workaround which is shown below.
``` python
self.update_state(
    state=states.FAILURE,
    meta={
        'exc_type': 'Conflict',
        'exc_module': 'werkzeug.exceptions',
        'exc_message': 'This token already exists'
    }
)
```

**Comment by auvipy** [comment]
Hey, Ava thanks for your detail report. I am not sure celery directly serialize werkzeug errors. may be you are using flask? can you provide a test script which we can use to reproduce?

**Reply by amthorn** [reply]
Here's a test script that can reliably reproduce (using redis as a backend and rabbitmq as a broker):

** app.py **
``` python
from celery import Celery
import werkzeug

app = Celery(
    'tasks',
    backend='redis://redis:6379/0',
    broker='pyamqp://guest@rabbitmq:5672//'
)

@app.task(bind=True)
def myTask(self):
    raise werkzeug.exceptions.NotFound("This message won't be propogated")
```

Run the worker with `celery --app app worker`

** test.py **
``` python
from app import myTask

result = myTask.delay()
print(result.info)
```

When you run test.py, you will notice that the error message string is not propogaged and the output of the script will simply just show what is shown below:
```
404 Not Found: The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.
```

When i expect to see:
```
404 Not Found: This message won't be propogated
```

I can also confirm that in redis when i retrieve the key using the redis CLI, the error message is not properly propagated as shown below:
```
127.0.0.1:6379> get celery-task-meta-ec6bc9e2-76dd-4c9a-bc1e-66b8d9f71c3e
"{\"status\": \"FAILURE\", \"result\": {\"exc_type\": \"NotFound\", \"exc_message\": [], \"exc_module\": \"werkzeug.exceptions\"}, \"traceback\": \"Traceback (most recent call last):\\n  File \\\"/usr/local/lib/python3.10/site-packages/celery/app/trace.py\\\", line 451, in trace_task\\n    R = retval = fun(*args, **kwargs)\\n  File \\\"/usr/local/lib/python3.10/site-packages/celery/app/trace.py\\\", line 734, in __protected_call__\\n    return self.run(*args, **kwargs)\\n  File \\\"/src/app.py\\\", line 16, in myTask\\n    raise werkzeug.exceptions.NotFound(\\\"This message won't be propogated\\\")\\nwerkzeug.exceptions.NotFound: 404 Not Found: This message won't be propogated\\n\", \"children\": [], \"date_done\": \"2021-12-12T11:01:56.522079\", \"task_id\": \"ec6bc9e2-76dd-4c9a-bc1e-66b8d9f71c3e\"}"
127.0.0.1:6379>
```
**NOTE**: the exc_message key is set to the empty list when it should be a list containing the string error message "This message won't be propagated". If i change the error message to a non-werkzeug exception, such as a type error. Then redis will show the proper setting for exc_message as shown below:
```
127.0.0.1:6379> get celery-task-meta-09564e77-02d4-40e3-960e-cae202a58496
"{\"status\": \"FAILURE\", \"result\": {\"exc_type\": \"TypeError\", \"exc_message\": [\"This message won't be propogated\"], \"exc_module\": \"builtins\"}, \"traceback\": \"Traceback (most recent call last):\\n  File \\\"/usr/local/lib/python3.10/site-packages/celery/app/trace.py\\\", line 451, in trace_task\\n    R = retval = fun(*args, **kwargs)\\n  File \\\"/usr/local/lib/python3.10/site-packages/celery/app/trace.py\\\", line 734, in __protected_call__\\n    return self.run(*args, **kwargs)\\n  File \\\"/src/app.py\\\", line 16, in myTask\\n    raise TypeError(\\\"This message won't be propogated\\\")\\nTypeError: This message won't be propogated\\n\", \"children\": [], \"date_done\": \"2021-12-12T11:07:07.370749\", \"task_id\": \"09564e77-02d4-40e3-960e-cae202a58496\"}"
```

In addition, when using a type error, the error is propagated appropriately back up into the task as a task result.

TLDR: it works with a TypeError and then, changing nothing else but what kind of exception is being raised, it doesn't work with a werkzeug exception

**Reply by auvipy** [reply]
I am not sure celery can serialize werkzeug exceptions, can you try flask framework instead? And see if celery is working with that

**Reply by amthorn** [reply]
I already have a work around (I posted at the bottom of the OP) so I’m not necessarily looking for a way to achieve what I need. I simply wanted to start this discussion to see if this is a bug in celery or not.

If the intent is that celery should not propagate all error messages then we could conclude this discussion
